jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/GracefulShutdownHandler.java

```diff
@@ -0,0 +1,143 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.server.handler;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.LongAdder;
+
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.component.Graceful;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Handler to track active requests and allow them to gracefully complete.
+ */
+public class GracefulShutdownHandler extends Handler.Wrapper implements Graceful
+{
+    private static final Logger LOG = LoggerFactory.getLogger(GracefulShutdownHandler.class);
+
+    private final LongAdder dispatchedStats = new LongAdder();
+    private final Shutdown shutdown;
+
+    public GracefulShutdownHandler()
+    {
+        shutdown = new Shutdown(this)
+        {
+            @Override
+            public boolean isShutdownDone()
+            {
+                long count = dispatchedStats.sum();
+                if (LOG.isDebugEnabled())
+                    LOG.debug("isShutdownDone: count {}", count);
+                return count == 0;
+            }
+        };
+    }
+
+    /**
+     * Flag indicating that Graceful shutdown has been initiated.
+     *
+     * @return whether the graceful shutdown has been initiated
+     * @see Graceful
+     */
+    @Override
+    public boolean isShutdown()
+    {
+        return shutdown.isShutdown();
+    }
+
+    @Override
+    public boolean process(Request request, Response response, Callback callback) throws Exception
+    {
+        // Increment the counter before the test for isShutdown(), to avoid race conditions.
+        ShutdownTrackingCallback shutdownCallback = new ShutdownTrackingCallback(request, response, callback);
+
+        Handler handler = getHandler();
+        if (handler == null || !isStarted() || isShutdown())
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Service Unavailable: {}", request.getHttpURI());
+            Response.writeError(request, response, shutdownCallback, HttpStatus.SERVICE_UNAVAILABLE_503);
+            return true;
+        }
```

Comment:
This still does not feel correct.  If we are not started, then we should not be processing the request, even for a 503. 

```suggestion
        Handler handler = getHandler();
        if (handler == null || !isStarted())
            return false;
        
        // Increment the counter before the test for isShutdown(), to avoid race conditions.
        ShutdownTrackingCallback shutdownCallback = new ShutdownTrackingCallback(request, response, callback);

        if (isShutdown())
        {
            if (LOG.isDebugEnabled())
                LOG.debug("Service Unavailable: {}", request.getHttpURI());
            Response.writeError(request, response, shutdownCallback, HttpStatus.SERVICE_UNAVAILABLE_503);
            return true;
        }
```

The `handler==null` case can be before or after.

Reply:
Changed.